File: src/main/java/dev/goerner/geozen/model/multi_geometry/MultiPolygon.java
package dev.goerner.geozen.model.multi_geometry;

import dev.goerner.geozen.model.CoordinateReferenceSystem;
import dev.goerner.geozen.model.Geometry;
import dev.goerner.geozen.model.Position;
import dev.goerner.geozen.model.simple_geometry.Polygon;

import java.util.ArrayList;
import java.util.List;

/**
 * A {@link MultiPolygon} is a {@link Geometry} that represents a collection of {@link Polygon Polygons} in space. It is
 * defined by a list of {@link Polygon Polygons} and a {@link CoordinateReferenceSystem}.
 */
public class MultiPolygon extends Geometry {

    private final List<List<List<Position>>> coordinates;

    /**
     * Creates a new {@link MultiPolygon} with the given list of {@link Polygon Polygons} and the default WGS 84
     * {@link CoordinateReferenceSystem}.
     *
     * @param coordinates The list of {@link Polygon Polygons} representing the {@link MultiPolygon}.
     */
    public MultiPolygon(List<List<List<Position>>> coordinates) {
        this(coordinates, CoordinateReferenceSystem.WGS_84);
    }

    /**
     * Creates a new {@link MultiPolygon} with the given list of {@link Polygon Polygons} and the given
     * {@link CoordinateReferenceSystem}.
     *
     * @param coordinates               The list of {@link Polygon Polygons} representing the {@link MultiPolygon}.
     * @param coordinateReferenceSystem The {@link CoordinateReferenceSystem} of the {@link MultiPolygon}.
     */
    public MultiPolygon(List<List<List<Position>>> coordinates, CoordinateReferenceSystem coordinateReferenceSystem) {
        super(coordinateReferenceSystem);
        List<List<List<Position>>> coordsCopy = new ArrayList<>();
        for (List<List<Position>> polygon : coordinates) {
            List<List<Position>> polygonCopy = new ArrayList<>();
            for (List<Position> ring : polygon) {
                polygonCopy.add(List.copyOf(ring));
            }
            coordsCopy.add(List.copyOf(polygonCopy));
        }
        this.coordinates = List.copyOf(coordsCopy);
    }

    @Override
    public double getFastDistanceTo(Geometry other) {
        throw new UnsupportedOperationException("Fast distance calculation not implemented yet");
    }

    @Override
    public double getExactDistanceTo(Geometry other) {
        throw new UnsupportedOperationException("Exact distance calculation not implemented yet");
    }

    public List<List<List<Position>>> getCoordinates() {
        return coordinates;
    }
}

File: src/main/kotlin/dev/goerner/geozen/model/multi_geometry/MultiPolygon.kt
package dev.goerner.geozen.model.multi_geometry

import dev.goerner.geozen.model.CoordinateReferenceSystem
import dev.goerner.geozen.model.Geometry
import dev.goerner.geozen.model.Position

/**
 * A [MultiPolygon] is a [Geometry] that represents a collection of [Polygons][dev.goerner.geozen.model.simple_geometry.Polygon] in space. It is
 * defined by a list of [Polygons][dev.goerner.geozen.model.simple_geometry.Polygon] and a [CoordinateReferenceSystem].
 *
 * @param coordinates The list of polygons (each represented as a list of rings, where each ring is a list of positions).
 *                    The outer list may be empty. Each polygon must have at least 1 ring (exterior ring).
 *                    Each ring must contain at least 4 positions and be closed (first position equals last position).
 * @param coordinateReferenceSystem The coordinate reference system, defaults to WGS_84
 * @throws IllegalArgumentException if any polygon has no rings, if any ring has fewer than 4 positions,
 *                                  or if any ring is not closed (first position != last position)
 */
data class MultiPolygon(
    val coordinates: List<List<List<Position>>>,
    override val coordinateReferenceSystem: CoordinateReferenceSystem = CoordinateReferenceSystem.WGS_84
) : Geometry(coordinateReferenceSystem) {

    init {
        coordinates.forEachIndexed { polygonIndex, polygon ->
            require(polygon.isNotEmpty()) {
                "Each Polygon in MultiPolygon must contain at least 1 ring (exterior ring), " +
                        "but Polygon at index $polygonIndex contained 0"
            }

            polygon.forEachIndexed { ringIndex, ring ->
                require(ring.size >= 4) {
                    val ringType = if (ringIndex == 0) "exterior ring" else "interior ring at index $ringIndex"
                    "Each ring in MultiPolygon must contain at least 4 positions, " +
                            "but Polygon at index $polygonIndex $ringType contained ${ring.size}"
                }

                require(ring.first() == ring.last()) {
                    val ringType = if (ringIndex == 0) "exterior ring" else "interior ring at index $ringIndex"
                    "Each ring in MultiPolygon must be closed (first position must equal last position), " +
                            "but Polygon at index $polygonIndex $ringType is not closed"
                }
            }
        }
    }

    override fun fastDistanceTo(other: Geometry): Double {
        throw UnsupportedOperationException("Fast distance calculation not implemented yet")
    }

    override fun exactDistanceTo(other: Geometry): Double {
        throw UnsupportedOperationException("Exact distance calculation not implemented yet")
    }
}
